### PR TITLE
Docker support

### PR DESCRIPTION
**Pull request by skiunke**
I added support for docker and docker-compose to wrap the ruby and database. Had to bump the ruby version from 3.2.2 -> 3.2.3 to make it work.

### CODE REVIEW

**Comment by kphurley** [comment]
Hey, sorry I kinda lost track of this.  I'll get to this sometime today.

File: Gemfile [hunk 1]
```diff
@@ -1,7 +1,8 @@
 source "https://rubygems.org"
 git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 
-ruby "3.2.2"
+ruby '3.2.3'
```
Comment (kphurley) [inline Gemfile:4]: This is a bit of a blocker for me at the moment.  In my local development flow, I'm using `rvm` on osx, and it wasn't able to switch to this version of ruby cleanly.  I'll probably have to remove rvm and use homebrew to get to this version - and right now I'm prioritizing getting a release out, I have limited time to work and I'd rather spend it getting the app moving.

I'm probably going to sit on this change for now until the app is further along and I can release it.  I definitely think there is value in getting these things wrapped in containers, but I don't want to do it at the expense of getting an MVP out.

Comment (esabol) [reply Gemfile:4]: Question for @skiunke : Why does the Docker container require 3.2.3? If 3.2.2 works for @kphurley outside of Docker, then it should be possible to get 3.2.2 working inside the Docker container. Just my two cents!

Comment (skiunke) [reply Gemfile]: @kphurley @esabol I pushed a fix reverting to 3.2.2.
The original error message was somehow cryptic but specifying the version more clearly in the Dockerfile fixed the issue.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,34 @@
+FROM ruby:3.2.2
+
+RUN curl -sL https://deb.nodesource.com/setup_20.x | bash - && \
```
Comment (esabol) [inline Dockerfile:3]: It seems `curl -sL https://deb.nodesource.com/setup_20.x | bash -` is deprecated, according to https://github.com/nodesource/distributions/discussions/1639. The new, approved method involves a bunch more commands. You can find it at that URL in multiple answers, but it's something like this:
```
# Install NodeJs
ARG NODE_MAJOR=20
RUN apt-get update && \
   apt-get install -y ca-certificates curl gnupg && \
   curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor -o /etc/apt/trusted.gpg.d/nodesource.gpg && \
   echo "deb [signed-by=/etc/apt/trusted.gpg.d/nodesource.gpg] https://deb.nodesource.com/node_$NODE_MAJOR.x nodistro main" | tee /etc/apt/sources.list.d/nodesource.list && \
   apt-get update && \
   apt-get install -y nodejs
# End Install
```
I have tried it. What do you think?

File: Gemfile.lock [hunk 3]
```diff
@@ -213,7 +214,7 @@ DEPENDENCIES
   web-console
 
 RUBY VERSION
-   ruby 3.2.2p53
+   ruby 3.2.3p157
```
Comment (esabol) [inline Gemfile.lock:217]: I'm not a Ruby expert, so I'm not sure this matters, but this line still references 3.2.3?

File: README.md [hunk 4]
```diff
@@ -5,16 +5,25 @@ This is a deck building app for Star Wars: The Card Game.  This project is inspi
...
 ### You need
-- Ruby 3.2
-- Rails 7
-- Postgres 14
+
+1. For local deployment
+    - Ruby 3.2
+    - Rails 7
+    - Postgres 14
+2. For docker
```
Comment (esabol) [inline README.md:13]: Maybe `For container deployment` here to avoid the redundancy with the next line?

File: Gemfile [hunk 5]
```diff
@@ -1,7 +1,8 @@
 source "https://rubygems.org"
 git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 
-ruby "3.2.2"
+ruby '3.2.2'
+
```
Comment (esabol) [inline Gemfile:5]: This change to the Gemfile seems unnecessary?

File: Gemfile.lock (hunk 3 above)
Comment (skiunke) [reply Gemfile.lock:217]: forgot to push it it got reverted to 3.2.2p53

File: README.md (hunk 4 above)
Comment (skiunke) [reply README.md:13]: fixed

File: Dockerfile (hunk 2 above)
Comment (skiunke) [reply Dockerfile:3]: i didn't get the outdated warning in CI, but i changed it to the recommended installation way

**Comment by skiunke** [comment]
@kphurley pushed the fixed to your comments

File: Dockerfile [hunk 6]
```diff
@@ -0,0 +1,29 @@
+FROM ruby:3.2.2
+
+RUN apt-get update  && \
+    apt-get install -y ca-certificates curl gnupg  &&\
```
Comment (esabol) [inline Dockerfile:4]: Sorry for the nitpicking, but add a space between the && and \.

File: Dockerfile [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+    mkdir -p /etc/apt/keyrings && \
+    curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor -o /etc/apt/keyrings/nodesource.gpg && \
+    NODE_MAJOR=20  && \
```
Comment (esabol) [inline Dockerfile:7]: Remove one of the excess spaces before &&.

File: Dockerfile [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_$NODE_MAJOR.x nodistro main" > /etc/apt/sources.list.d/nodesource.list && \
+    apt-get update  &&\
```
Comment (esabol) [inline Dockerfile:9]: Remove one of the excess spaces before && and add a space after the &&.

**Review by esabol — CHANGES_REQUESTED** [review]
Just some last trivial whitespace changes. Sorry, I'm a bit of a perfectionist when it comes to whitespace.

File: Dockerfile (hunk 7 above)
Comment (esabol) [reply Dockerfile:7]: Also, you didn't like my version in which NODE_MAJOR was an ARG?

Comment (skiunke) [reply Dockerfile:7]: I just copied the code from the website

**Comment by skiunke** [comment]
@esabol i fixed the whitespaces @kphurley I renamed the env variables for docker to be used in production and updated the readme again to be more clearly.

File: Dockerfile (hunk 7 above)
Comment (esabol) [reply Dockerfile]: Well, you could have copied the code from my comment. :)

But whatever. It's not that important.

**Review by esabol — APPROVED** [review]

File: postgres_docker.sh [hunk 9]
```diff
@@ -0,0 +1 @@
+docker run --name postgres-swlcgdb -e POSTGRES_DB=swlcgdb_development -e POSTGRES_USER=swlcgdb -e POSTGRES_PASSWORD=swlcgdb_development -p 5432:5432 -d postgres:16
```
Comment (esabol) [inline postgres_docker.sh:1]: You need a carriage return (or linefeed, whatever you want to call it)  at the end of this line. It's a POSIX requirement.

https://stackoverflow.com/a/729795/5153779

File: config/database.yml [hunk 10]
```diff
@@ -81,6 +86,8 @@ test:
...
 production:
   <<: *default
-  database: swlcgdb_production
-  username: swlcgdb
-  password: <%= ENV["SWLCGDB_DATABASE_PASSWORD"] %>
+  database: <%= ENV["DATABASE_NAME"] %>
+  username: <%= ENV["DATABASE_USERNAME"] %>
+  password: <%= ENV["DATABASE_PASSWORD"] %>
+  host: <%= ENV["DATABASE_HOST"] %>
+  port: <%= ENV["DATABASE_PORT"] %>
```
Comment (esabol) [inline config/database.yml:93]: Add a carriage return to the end of this file as well.

**Review by esabol — CHANGES_REQUESTED** [review]

File: README.md [hunk 11]
```diff
@@ -3,18 +3,38 @@
...
+    - Docker
+    - Docker Compose
+
+### For Development
+1. `git clone https://github.com/kphurley/swlcgdb.git swlcgdb`
+2. `cd swlcgdb`
+3. Start postgres
+   4. Configure config/database.yml if required
+   5. To start a docker instance of postgres run `./postgres_docker.sh`
+4. `npx install esbuild`
```
Comment (kphurley) [inline README.md:23]: Small nit - This markdown renders a little oddly - mind cleaning that up?

It's probably ok here to have one long series of steps with (optional) in front of them.  I think the markdown renderer gets confused with this indentation.

File: config/database.yml (hunk 10 above)
Comment (kphurley) [reply config/database.yml:93]: +1

File: postgres_docker.sh (hunk 9 above)
Comment (kphurley) [reply postgres_docker.sh:1]: +1

**Review by kphurley — CHANGES_REQUESTED** [review]
I think this is pretty close now.  Thanks for getting this working with my current version of Ruby.  I was able to get the docker process rolling with one additional change - since we made a change to `package.json`, the lockfile (`package-lock.json`) should also be committed.  Otherwise the `npm ci` command initializing Docker will fail.

**Review by esabol — APPROVED** [review]

**Review by kphurley — COMMENTED** [review]
This is looking good.  I'll give this an end to end spin tonight and hopefully approve after that.

**Review by kphurley — APPROVED** [review]

**Pull request merged**